lib/pymedphys/_experimental/streamlit/apps/sum_doses.py

```diff
@@ -0,0 +1,318 @@
+) -> "pydicom.dataset.Dataset":
+    """Sum two or more DICOM dose grids and save to new DICOM RT
+    Dose dataset"
+
+    Parameters
+    ----------
+    datasets : sequence of pydicom.dataset.Dataset
+        A sequence of DICOM RT Dose datasets whose doses are to be
+        summed.
+
+    Returns
+    -------
+    pydicom.dataset.Dataset
+        A new DICOM RT Dose dataset whose dose is the sum of all doses
+        within `datasets`
+    """
+
+    if not len(datasets) >= 2:
+        raise ValueError("At least two datasets must be provided for comparison")
+
+    if not all(ds.Modality == "RTDOSE" for ds in datasets):
+        raise ValueError("`datasets` must only contain DICOM RT Dose datasets.")
+
+    if not patient_ids_in_datasets_are_equal(datasets):
+        raise ValueError("Patient ID must match for all datasets")
+
+    if not all(ds.DoseSummationType == "PLAN" for ds in datasets):
+        raise ValueError(
+            "Only DICOM RT Doses whose DoseSummationTypes are 'PLAN' are supported"
+        )
+
+    if not all(ds.DoseUnits == datasets[0].DoseUnits for ds in datasets):
+        raise ValueError(
+            "All DICOM RT Doses must have the same units ('GY or 'RELATIVE')"
+        )
+
+    if not coords_in_datasets_are_equal(datasets):
+        raise ValueError("All dose grids must have perfectly coincident coordinates")
+
+    ds_summed = copy.deepcopy(datasets[0])
+
+    ds_summed.BitsAllocated = 32
+    ds_summed.BitsStored = 32
+    ds_summed.DoseSummationType = "MULTI_PLAN"
+    ds_summed.DoseComment = "Summed Dose"
+
+    if not all(ds.DoseType in ("PHYSICAL", "EFFECTIVE") for ds in datasets):
+        raise ValueError(
+            "Only DICOM RT Doses whose DoseTypes are 'PHYSICAL' or "
+            "'EFFECTIVE' are supported"
+        )
+    if any(ds.DoseType == "EFFECTIVE" for ds in datasets):
+        ds_summed.DoseType = "EFFECTIVE"
+    else:
+        ds_summed.DoseType = "PHYSICAL"
+    doses = np.array([dose_from_dataset(ds) for ds in datasets])
+    doses_summed = np.sum(doses, axis=0, dtype=np.float32)
+
+    # ds_summed.DoseGridScaling = np.max(doses_summed) / (2 ^ int(ds_summed.HighBit))
+
+    ds_summed.DoseGridScaling = np.max(doses_summed) / (2 ** 32 - 1)
+    pixel_array_summed = (doses_summed / ds_summed.DoseGridScaling).astype(np.uint32)
+
+    ds_summed.PixelData = pixel_array_summed.tobytes()
+
+    return ds_summed
+
+
+def main():
+
+    left_column, right_column = st.beta_columns(2)
+
+    with left_column:
+        st.write("## Upload DICOM RT Dose files")
+        files = st.file_uploader(
+            "Upload at least two DICOM RT Dose files whose doses you'd "
+            "like to add together. The first file uploaded will be "
+            "used as a template for the summed DICOM RT Dose file.",
+            ["dcm"],
+            accept_multiple_files=True,
+        )
+
+    if st.button("Sum Doses"):
+
+        datasets = _check_files_valid(files)
+
+        if datasets:
```

Comment:
```suggestion
        if not datasets:
	        st.stop()
```

And then all of the below need not be nested. The app will stop at this line if no datasets are uploaded.

Maybe instead have:

```python
if not files:
    st.stop()
```

Right below the `st.file_uploader` line

Reply:
Yeah, I wondered about the best way to do all of this!

If I go for `if not files`, I'll still either need to keep `if not datasets:` or have `_check_files_valid()` raise errors rather than returning `None`. I initially wanted to avoid `_check_files_valid()` raising errors to avoid the user having to re-upload all files, since I thought that Streamlit would crash when such an error is thrown. Now I've learned that Streamlit has lovely inbuilt error handling, so I can just throw the errors after all...

Reply:
So, pyright is hyper picky about this. Unnesting gives this:

![image](https://user-images.githubusercontent.com/19767103/106714440-c514ff00-664b-11eb-8ba5-e2c8f617570f.png)

So, I can either re-nest, or "optionalise" the type hint for `sum_doses_in_datasets()`.

Nesting seems cleaner...

Reply:
Would you be okay if I did a bit of refactoring?

Reply:
More than happy!

Reply:
:slightly_smiling_face: